Display serialised tx endpoint

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->

adp-974

- [x] swagger update
- [x] adding types
- [x] adding `decodeTransaction` in Server
- [x] add `lookupTxIns` in Wallet
- [x] add `lookupTxOuts` in Wallet
- [x] adjust unit testing
- [x] add integration tests for inputs/outputs/withdrawals/metadata

The endpoint acts in the context of a given wallet and is able to detect the wallet's resources. In detail, the enpdoint detects:
- id of tx
- fee
- inputs, belonging to other wallets and the wallets - in that case it lookups utxo and give amount and derivation path
- outputs, belonging to other wallets or to the wallets - in that case it either see its output or change (here it looks ahead using gap) and enrich info by current amount and derivation path of address
- withdrawals, belonging to other wallet or the wallet
- metadata
- script validity

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

paweljakubas committed:
add first version of endpoint in swagger

paweljakubas committed:
add basic types

paweljakubas committed:
change inputs to be more general in ApiDecodedTransaction

clean

paweljakubas committed:
deal with colls, inps and outs

paweljakubas committed:
deal with withdrawals

paweljakubas committed:
deal with validity interval

paweljakubas committed:
add lookupTxIns in Cardano.Wallet

paweljakubas committed:
use lookupTxIns for inps and collaterals in decodeTransaction

paweljakubas committed:
improve handling inputs

paweljakubas committed:
generalize withdrawals in swagger

paweljakubas committed:
deal with withdrawals

paweljakubas committed:
first integration test for decode tx

paweljakubas committed:
rm unneeded errors from ErrBalanceTx

paweljakubas committed:
unpend single output tx test

paweljakubas committed:
fix address comparing in lookupTxIns

paweljakubas committed:
finish integration test for inputs

paweljakubas committed:
amend ExternalInput definition - we do not know amount from cbor here

paweljakubas committed:
generalize outputs in swagger

paweljakubas committed:
add generalized output type

paweljakubas committed:
first impl of txout enricher

paweljakubas committed:
add more tests for outputs of wallet in integration testing

post rebase adjustments

paweljakubas committed:
add change addr detection

some cleaning

paweljakubas committed:
use pool gap to determine number of candidate change addresses

paweljakubas committed:
add integration test for metadata

paweljakubas committed:
add integration testing for withdrawals

paweljakubas committed:
adjust core unit tests

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1149,6 +1154,128 @@ manageRewardBalance _ ctx wid = db & \DBLayer{..} -> do
+                    [] -> (txin, Nothing) -- will probably use error here
+                    path':_ -> (txin, Just (txout, path'))
+
+extAddr
```
1. We could probably find a more descriptive name than `extAddr`. Seems similar to the existing `normalizeDelegationAddress`, but perhaps different enough not to combine.
2. **But**... Why are we modifying the address in the first place? 🤔 That seems very off to me. We should be looking up addresses based on their payment key hashes only, right?

paweljakubas replied:
ad 1. yes, `normaliseDelegationAddress` does almost the same as `extAddr`
ad 2. Why we needed? Because in the cbor there is delegation address as represented in the chain. and we store fingerprints really, so we need to adjust them in order to compare.

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1149,6 +1154,128 @@ manageRewardBalance _ ctx wid = db & \DBLayer{..} -> do
+    -> XPub
+    -> Address
+extAddr addr xpub =
+    liftDelegationAddress @n @k ((\(Right finger) -> finger) $
```
Some `either (error "extAddr: impossible: Address cannot be malformed")` would be nicer than just ignoring the `Left` case here.

paweljakubas replied:
I will rely on `normaliseDelegationAddress` rather than this

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1149,6 +1154,128 @@ manageRewardBalance _ ctx wid = db & \DBLayer{..} -> do
+  where
+    db = ctx ^. dbLayer @IO @s @k
+    tryGetTxOutPath cp walletAddrs txin =
+        case UTxO.lookup txin (totalUTxO mempty cp) of
```
Isn't `totalUTxO mempty cp` the same thing as `utxo cp`? But seems good that we don't remove the inputs from the pending txs — I take the pending txs wouldn't be decodable/displayable then.

Also, not sure we _need to_ nor if we _should_, but we _could_:
- Use `changeUTxO pending s <> utxo` (but not excluding pending inputs), to be able to resolve chains of dependent txs (I don't think this would work currently)
- Use LSQ to (potentially) `lookup` the `txin` in the node's full `UTxO`

paweljakubas replied:
I was thinking about pending txs. I would for now use `totalUTxO mempty cp` and think about pending tx in the next move

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1149,6 +1154,128 @@ manageRewardBalance _ ctx wid = db & \DBLayer{..} -> do
+                let path = map (\(_, _,p) -> p) $
+                        filter (\(addr', _,_) -> addr' == addr) walletAddrs
+                in case path of
+                    [] -> (txin, Nothing) -- will probably use error here
```
> will probably use error here

Sounds sane. Interesting case 😄 — yes, if the TxIn can be looked up from the utxo, it should exist in `knownAddresses`, I would presume. Unless we were to use some different notion of `utxo` (see the other comment).

Anviking replied:
> Unless we were to use some different notion of utxo

Or maybe for this reason keep it as is, just in case 🤷‍♂️

paweljakubas replied:
yes, maybe for now let's keep it as it is

paweljakubas committed:
fixing schema/json things

paweljakubas committed:
use normaliseDelegationAddress

Anviking committed:
Simplify lookupTx{Ins,Outs} using isOurs

This commit does sacrifice the distinction between amount and
amount_incoming. The next commit addresses that.

paweljakubas commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1171,103 +1165,37 @@ lookupTxIns ctx wid txins = db & \DBLayer{..} -> do
-            , ix )
-    let walletAddrs = map f $ knownAddresses (getState cp)
-
-    --We are analyzing the next gapPool change addresses
```
so we need this changeAddrs, otherwise source wallet will not see change address in outputs UNTIL the wallet discovers tx (so after submit). That is consequence of decoupling construction/signing/submitting and persisting pending things only during submitting

paweljakubas commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -479,7 +479,7 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
         let addrDest = (addrs !! addrIx) ^. #id
         let expectedTxOutTarget = WalletOutput $ ApiWalletOutput
                 { address = addrDest
-                , amount = Quantity 0
```
no it should be 0. This is what is currently in this address. So after construction, source wallet does not see this as WalletOutput but ExternalOutput. Target wallet see this as WalletOuptut, and see `amountIncoming=Quantity amt`, but `amount=Quantity 0` as there is nothing at THIS stage in this wallet. Right? It is going to be `amount=amountIncoming` after discovery. I did this on purpose to have a way to see whole cycle if this resource is ours :-)

paweljakubas replied:
see line 548 and 
```
        let expectedTxOutTarget' = WalletOutput $ ApiWalletOutput
                { address = addrDest
                , amount = Quantity amt -- now we have this
                , assets = ApiT TokenMap.empty
                , derivationPath = NE.fromList
                    [ ApiT (DerivationIndex 2147485500)
                    , ApiT (DerivationIndex 2147485463)
                    , ApiT (DerivationIndex 2147483648)
                    , ApiT (DerivationIndex 0)
                    , ApiT (DerivationIndex $ fromIntegral addrIx)
                    ]
                , amountIncoming = Quantity amt
                , assetsIncoming = ApiT TokenMap.empty
                }
```

paweljakubas replied:
this is after submitting and discovering

Anviking committed:
Remove distinction between amount and amount_incoming

My reason being the combination of:

1. Unclear why we need it for this task / now. There might be interesting things to do here,
   but that we should be clear about what problem we're trying to solve, such that we can create
   a solution we're fully confident in.
2. Exposing balances of individual addresses feels at odds with the point of the wallet otherwise
   being maintaining a concept of balance for all addresses combined.
3. It seemed to make the implementation much more complicated

Anviking committed:
Tweak integration expectations

1. Check that the change output is recognized
2. Use `shouldContain` and `shouldNotContain` rather than
`shouldSatisfy` and `elem` for nicer error messages.

Anviking reviewed (approved):
lgtm! Hoping we at some point later on can get CBOR <-> JSON roundtrip properties, which I believe would be the ultimate test that we display all information!

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -395,8 +456,62 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
+            , expectField #outputs (`shouldNotContain` [expectedTxOutTarget])
+
+            -- Check that the change output is there:
+            , expectField (#outputs) ((`shouldBe` 1) . length . filter isOutOurs)
```
I added a check for the change output in the last commit along with a slight tweak, using `should{Not}Contain [txOut]` rather than `shouldSatisfy` and `elem`, which should produce nicer error messages and be slightly simpler.

paweljakubas replied:
good change
